File: queue.rb
load "#{Dir.home}/Documents/Interviews/algo-practice/uni_node.rb"

class Queue
  attr_reader :length

  def initialize
    @first = nil
    @last = nil
    @length = 0
  end

  def push(data)
    new_last = UniNode.new(data)
    @last.next_node = new_last unless @last.nil?
    @last = new_last
    @first = new_last if @first.nil?
    @length += 1
  end

  def pop
    raise "EMPTY QUEUE" if is_empty?
    data_to_pop = @first.data
    @first = @first.next_node
    @last = nil if @first.nil?
    @length -= 1
    data_to_pop
  end

  def is_empty?
    @first.nil?
  end

  def peek
    raise "EMPTY QUEUE" if is_empty?
    return @first.data
  end    
end

def test
  q = Queue.new
  q.push(5)
  puts q.peek == 5 ? 'PASS' : 'FAIL'
  puts q.is_empty? == false ? 'PASS' : 'FAIL'
  puts q.length == 1 ? 'PASS' : 'FAIL'

  q.push(4)
  puts q.peek == 5 ? 'PASS' : 'FAIL'
  puts q.is_empty? == false ? 'PASS' : 'FAIL'
  puts q.length == 2 ? 'PASS' : 'FAIL'

  puts q.pop == 5 ? 'PASS' : 'FAIL'
  puts q.is_empty? == false ? 'PASS' : 'FAIL'
  puts q.length == 1 ? 'PASS' : 'FAIL'

  puts q.pop == 4 ? 'PASS' : 'FAIL'
  puts q.is_empty? ? 'PASS' : 'FAIL'
  puts q.length == 0 ? 'PASS' : 'FAIL'
end

File: utils_python/queue.py
class QueueNode:
  def __init__(self, data, next_node = None):
    self.data = data
    self.next_node = next_node

class Queue:
  def __init__(self):
    self.first = None
    self.last = None

  def is_empty(self):
    return self.first == None

  def peek(self):
    if self.first == None:
      raise IndexError
    return self.first.data

  def add(self, data):
    n = QueueNode(data, self.last)
    if self.first == None:
      self.first, self.last = n, n
    else:
      self.last.next_node = n
      self.last = n
    return True

  def remove(self):
    if self.first == None:
      raise IndexError
    data_to_return = self.first.data
    self.first = self.first.next_node
    if self.first == None:
      self.last = None
    return data_to_return

def test():
  q = Queue()
  try:
    q.remove()
  except IndexError:
    pass

  try:
    q.peek()
  except IndexError:
    pass

  q.add(1)
  print('PASS' if q.peek() == 1 else 'FAIL')
  q.add(2)
  print('PASS' if q.peek() == 1 else 'FAIL')
  print('PASS' if q.remove() == 1 else 'FAIL')
  print('PASS' if q.peek() == 2 else 'FAIL')
  print('PASS' if q.remove() == 2 else 'FAIL')

  try:
    q.remove()
  except IndexError:
    pass

  try:
    q.peek()
  except IndexError:
    pass

if __name__ == '__main__':
  test()
